### ISSUE TITLE
problem with chaining a resolved promise

### ISSUE THREAD

**Issue by catmando** [opal/opal#1942]
I am not sure if this is bug or a misunderstanding of how promises work.

```ruby
  1) reactive-record edge cases chaining a resolved promise
     Failure/Error:
       expect_evaluate_ruby do
         x = Promise.new
         x.resolve
         y = x.then { Promise.new }
         z = y.then { |r| r }
         [x.resolved?, y.resolved?, z.resolved?]
       end.to eq([true, false, false])
     
       expected: [true, false, false]
            got: [true, true, false]
```

**Comment by catmando** [comment]
Changing https://github.com/hyperstack-org/opal/blob/6510ee2e14d8818afd8954c3d852da590e2623e6/stdlib/promise.rb#L206

to `resolve!(value) unless value.is_a? Promise` fixes the problem.

When I get some time I will put a spec together and do a PR.

**Comment by elia** [comment]
@meh do you know if that's the intended behavior?

**Comment by hmdne** [comment]
This should be fixed by #2220

**Comment by hmdne** [comment]
I checked and tried to understand the issue, thing is, it isn't fixed with PromiseV2 as of yet, as it behaves in this part exactly as Promise (contrary to how JS promises work), as in: if a Promise has been entered, it's marked as resolved.

Unfortunately, with PromiseV2 the resolve? call won't be reliable, as we can't inspect a JS promise. We only simulate the interface of PromiseV1 and for those cases it should mostly work.
